src/System.Device.Gpio/System/Device/Gpio/Drivers/Libgpiod/LibGpiodDriverFactory.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Device.Gpio.Drivers.Libgpiod.V1;
+using System.Device.Gpio.Drivers.Libgpiod.V2;
+using System.Device.Gpio.Libgpiod;
+using System.Device.Gpio.Libgpiod.V2;
+using System.IO;
+using System.Linq;
+
+namespace System.Device.Gpio.Drivers;
+
+/// <summary>
+/// Driver factory for different versions of libgpiod.
+/// </summary>
+internal static class LibGpiodDriverFactory
+{
+    private const string DriverVersionEnvVar = "DOTNET_IOT_LIBGPIOD_DRIVER_VERSION";
+    private const string LibrarySearchPattern = "libgpiod.so*";
+
+    private static readonly Dictionary<string, LibGpiodDriverVersion> _libraryToDriverVersionMap = new()
+    {
+        { "libgpiod.so.0", LibGpiodDriverVersion.V1 },
+        { "libgpiod.so.1", LibGpiodDriverVersion.V1 },
+        { "libgpiod.so.2", LibGpiodDriverVersion.V1 },
+        { "libgpiod.so.3", LibGpiodDriverVersion.V2 }
+    };
+
+    private static readonly Dictionary<LibGpiodDriverVersion, string[]> _driverVersionToLibrariesMap = new()
+    {
+        { LibGpiodDriverVersion.V1, new[] { "libgpiod.so.0", "libgpiod.so.1", "libgpiod.so.2" } },
+        { LibGpiodDriverVersion.V2, new[] { "libgpiod.so.3" } }
+    };
+
+    private static readonly LibGpiodDriverVersion[] _driverCandidates;
+
+    private static readonly string? _driverVersionEnvVarValue;
+    private static readonly LibGpiodDriverVersion? _driverVersionSetByEnvVar;
+
+    /// <remarks>
+    /// Driver version that is picked by this factory when no version is explicitly specified. Null when libpgiod is not installed.
+    /// </remarks>
+    private static readonly LibGpiodDriverVersion? _automaticallySelectedDriverVersion;
+
+    private static readonly string[] _librarySearchPaths = { "/lib", "/usr/lib", "/usr/local/lib" }; // Based on Linux FHS standard
+
+    static LibGpiodDriverFactory()
```

Comment:
In another comment there was a performance concern about the previous implementation that used a static method with a lock, so I moved this into a static constructor. I don't think we should move away from `static`, because executing this logic only once to cache the result makes sense for performance reasons. I do understand your concern though about the `TypeLoadException`, which is not really nice, so I would vote for the previous approach, using a static method instead.

Reply:
[Triage] We discussed this briefly and we think the best solution would be to merge the two concepts and make the _class_ a singleton. So have all the fields and methods non-static, but offer a static `Instance` property that creates that instance exactly once. Like this, the overhead of scanning is required only once, while any exceptions will still be deterministic and hence catchable. You'll find some hints about singleton implementation [here](https://csharpindepth.com/articles/singleton) (just in case). 

Sorry this takes so many rounds.

Reply:
No worries! I'm always open for improvements and learnings. It's true that the Singleton pattern can be an alternative to a static class, though I think it won't fully solve the problem, but only defer the root problem. 

```c#
class Factory
{
    private Factory()
    {
      if (someReason)
          throw new Exception();
    }

    public static Factory Instance { get; } = new Factory();
}
```

When accessing `Instance`, again a `TypeInitializationException` is thrown, for which there exists no sane reason.
I think the best solution here is to make sure not to throw when any static initializer is called.
I have combined your Singleton idea with my thoughts, and pushed changes that utilize `Lazy`. This prevents `TypeInitializationException` and only propagates e.g. PermissionDenied etc.

Reply:
Yes, the way you wrote it here is not safe about TypeInitialisationException. You can either use Lazy, or something llike this:

```csharp
public static Factory Instance
{
    get
    {
        lock (_lock)
        {
            if (_instance == null)
                _instance = new Factory();
        }
        return _instance;
    }
}
```

This way, the constructor is only called when the factory is accessed the first time and not when the class is accessed the first time. 

Note the difference between
```csharp
public static Factory Instance { get; } = new Factory();
```
and
```csharp
public static Factory Instance { get { return new Factory(); } };
```

These two look similar, but their semantics is very different: The first creates a static automatic property, which is initialized using an auto-generated static constructor, while the second creates a calculated property without a backing field.

Reply:
As mentioned in another comment it would be safe due to initializing only Lazy members, but your approach is more readable so let's go with that one.